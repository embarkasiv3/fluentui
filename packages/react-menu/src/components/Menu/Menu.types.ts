import * as React from 'react';
import { ObjectShorthandProps, ShorthandProps } from '@fluentui/react-utilities';
import { MenuListProps } from '../MenuList/index';
import { PositioningProps } from '@fluentui/react-positioning';

/**
 * Extends and drills down Menulist props to simplify API
 * {@docCategory Menu }
 */
export interface MenuProps extends MenuListProps {
  /**
   * Explicitly require children
   */

  children: React.ReactNode;
  /**
   * Whether the popup is open
   */
  open?: boolean;

  /**
   * Whether the popup is open by default
   */
  defaultOpen?: boolean;

  /**
   * Wrapper to style and add events for the popup
   */
  menuPopup?: ShorthandProps<React.HTMLAttributes<HTMLElement>>;

  /**
   * Where the menu is positioned with respect to the trigger
   */
  position?: PositioningProps['position'];

  /**
   * How the menu is aligned wtih respect to the trigger
   */
  align?: PositioningProps['align'];

  /*
   * Opens the menu on hover
   */
  onHover?: boolean;

  /**
   * Opens the menu on right click (context menu), removes all other menu open interactions
   */
  onContext?: boolean;
}

/**
 * {@docCategory Menu }
 */
export interface MenuState extends MenuProps {
  /**
   * Ref to the root slot
   */
  ref: React.MutableRefObject<HTMLElement>;

  /**
   * Whether the popup is open
   */
  open: boolean;

  /**
   * Callback to open/close the popup
   */
  setOpen: React.Dispatch<React.SetStateAction<boolean>>;

  /**
   * Internal react node that just simplifies handling children
   */
  menuList: React.ReactNode;

  /**
   * Internal react node that just simplifies handling children
   */
  menuTrigger: React.ReactNode;

  /**
   * Wrapper to style and add events for the popup
   */
  menuPopup: ObjectShorthandProps<React.HTMLAttributes<HTMLElement>>;

  /**
   * The ref for the popup
   */
  menuPopupRef: React.MutableRefObject<HTMLElement>;

  /**
   * The ref for the MenuTrigger, used for popup positioning
   */
  triggerRef: React.MutableRefObject<HTMLElement>;

  /**
   * Id for the MenuTrigger element for aria relationship
   */
  triggerId: string;
}
