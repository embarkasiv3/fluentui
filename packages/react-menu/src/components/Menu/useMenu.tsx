import * as React from 'react';
import {
  makeMergeProps,
  resolveShorthandProps,
  useMergedRefs,
  useControllableValue,
  useId,
  useOnClickOutside,
} from '@fluentui/react-utilities';
import { MenuProps, MenuState } from './Menu.types';
import { MenuTrigger } from '../MenuTrigger/index';
import { useFluent } from '@fluentui/react-provider';
import { getCode, keyboardKey } from '@fluentui/keyboard-key';

export const menuShorthandProps: (keyof MenuProps)[] = ['menuPopup'];

const mergeProps = makeMergeProps<MenuState>({ deepMerge: menuShorthandProps });

/**
 * Create the state required to render Menu.
 *
 * The returned state can be modified with hooks such as useMenuStyles,
 * before being passed to renderMenu.
 *
 * @param props - props from this instance of Menu
 * @param ref - reference to root HTMLElement of Menu
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory Menu }
 */
export const useMenu = (props: MenuProps, ref: React.Ref<HTMLElement>, defaultProps?: MenuProps): MenuState => {
  const { document } = useFluent();
  const triggerId = useId();

  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
      menuPopup: { as: 'div' },
      triggerId,
    },
    defaultProps,
    resolveShorthandProps(props, menuShorthandProps),
  );

  // TODO Better way to narrow types ?
  const children = React.Children.toArray(state.children) as React.ReactElement[];

  // TODO throw warnings in development safely
  if (children.length !== 2) {
    // eslint-disable-next-line no-console
    console.warn('Menu can only take one MenuTrigger and one MenuList as children');
  }

  // TODO use Popper for the trigger ref
  const triggerRef = React.useRef<HTMLElement>() as React.MutableRefObject<HTMLElement>;
  state.triggerRef = triggerRef;
  children.forEach(child => {
    if (child.type === MenuTrigger) {
      state.menuTrigger = child;
    } else {
      state.menuList = child;
    }
  });

  const [open, setOpen] = useControllableValue(state.open, state.defaultOpen);
  // TODO fix useControllableValue typing
  state.open = open !== undefined ? open : state.open;
  state.setOpen = React.useCallback(
    (...args) => {
      setOpen(...args);
    },
    [setOpen],
  );

  useMenuPopup(state);
  useOnClickOutside({ element: document, refs: [state.menuPopupRef, triggerRef], callback: () => setOpen(false) });
  return state;
};

const useMenuPopup = (state: MenuState) => {
  // TODO use Popper for the popup ref
  const menuPopupRef = React.useRef<HTMLElement>(null) as React.MutableRefObject<HTMLElement>;
  state.menuPopupRef = menuPopupRef;
  state.menuPopup.children = (Component, originalProps) => {
    const newProps = { 'aria-labelledby': state.triggerId, ...originalProps };

    if (state.onHover && !state.onContext) {
      newProps.onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
        state.setOpen(true);
        originalProps?.onMouseEnter?.(e);
      };
      newProps.onMouseLeave = (e: React.MouseEvent<HTMLElement>) => {
        state.setOpen(false);
        originalProps?.onMouseLeave?.(e);
      };
    }

    newProps.onBlur = (e: React.FocusEvent<HTMLElement>) => {
      state.setOpen(false);
      originalProps?.onBlur?.(e);
    };

    newProps.onKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
      originalProps?.onKeyDown?.(e);
      const keyCode = getCode(e);

      if (keyCode !== keyboardKey.Escape) {
        return;
      }

      state.setOpen(false);
    };

    return React.createElement(
      Component as React.ElementType,
      {
        ...newProps,
        ref: state.menuPopupRef,
      },
      state.menuList,
    );
  };

  return state;
};
