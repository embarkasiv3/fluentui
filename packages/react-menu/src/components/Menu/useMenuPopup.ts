import * as React from 'react';
import { getCode, keyboardKey, EnterKey, SpacebarKey } from '@fluentui/keyboard-key';
import { MenuState } from './Menu.types';
import { isOutsideMenu } from '../../utils/index';

interface UseMenuPopupState
  extends Pick<
    MenuState,
    | 'menuPopup'
    | 'menuPopupRef'
    | 'setOpen'
    | 'menuList'
    | 'onContext'
    | 'onHover'
    | 'triggerId'
    | 'triggerRef'
    | 'open'
    | 'isSubmenu'
  > {}

/**
 * A hook that sets the correct render of the menu popup slot through the children render function
 */
export const useMenuPopup = (state: UseMenuPopupState) => {
  const { menuPopup, menuPopupRef, setOpen, open, menuList, triggerRef, onHover, onContext, isSubmenu } = state;

  const dismissedWithKeyboardRef = React.useRef(false);
  React.useEffect(() => {
    if (dismissedWithKeyboardRef.current && !open) {
      triggerRef.current?.focus();
    }

    dismissedWithKeyboardRef.current = false;
  }, [triggerRef, dismissedWithKeyboardRef, open]);

  menuPopup.children = (Component, originalProps) => {
    const newProps = { role: 'presentation', ...originalProps };

    newProps.onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
      if (onHover && !onContext) {
        setOpen(true);
      }

      originalProps?.onMouseEnter?.(e);
    };

    newProps.onBlur = (e: React.FocusEvent<HTMLElement>) => {
      if (isOutsideMenu({ triggerRef, menuPopupRef, event: e })) {
        setOpen(false);
      }
      originalProps?.onBlur?.(e);
    };

    newProps.onKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
      const keyCode = getCode(e);
      if (keyCode === keyboardKey.Escape || (isSubmenu && keyCode === keyboardKey.ArrowLeft)) {
        setOpen(false);
        dismissedWithKeyboardRef.current = true;
        e.stopPropagation(); // Left and Escape should only close one menu at a time
      }

      // Dismiss is still handled by click, but keydown event should still be propagated
      // If a child uses keydown without a native click event, the menu will remain open
      // Only native click event will close
      if (keyCode === EnterKey || keyCode === SpacebarKey) {
        dismissedWithKeyboardRef.current = true;
      }

      originalProps?.onKeyDown?.(e);
    };

    // Assumption made that all clicks will close the popup if propagated from children
    // To stop clicks from closing the menu call stopPropagation
    newProps.onClick = (e: React.MouseEvent<HTMLElement>) => {
      setOpen(false);
      originalProps?.onClick?.(e);
    };

    return React.createElement(
      Component as React.ElementType,
      {
        ...newProps,
        ref: menuPopupRef,
      },
      menuList,
    );
  };

  return state as MenuState;
};
