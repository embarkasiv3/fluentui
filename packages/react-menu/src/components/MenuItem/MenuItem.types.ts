import * as React from 'react';
import { ComponentProps, ShorthandProps, ObjectShorthandProps } from '@fluentui/react-utilities';

export interface MenuItemProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * Icon slot rendered before children content
   */
  icon?: ShorthandProps<React.HTMLAttributes<HTMLElement>>;
}

export interface MenuItemState extends MenuItemProps {
  /**
   * Ref to the root slot
   */
  ref: React.MutableRefObject<HTMLElement>;
  /**
   * Icon slot when processed by internal state
   */
  icon?: ObjectShorthandProps<React.HTMLAttributes<HTMLSpanElement>>;
}
