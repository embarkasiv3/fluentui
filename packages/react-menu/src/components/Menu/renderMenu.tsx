import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { MenuState } from './Menu.types';
import { menuShorthandProps } from './useMenu';
import { MenuProvider } from '../../menuContext';

/**
 * Render the final JSX of Menu
 * {@docCategory Menu }
 */
export const renderMenu = (state: MenuState) => {
  const { slots, slotProps } = getSlots(state, menuShorthandProps);
  const {
    open,
    setOpen,
    onCheckedValueChange,
    checkedValues,
    defaultCheckedValues,
    onHover,
    onContext,
    triggerRef,
    triggerId,
  } = state;

  return (
    <MenuProvider
      value={{
        open,
        setOpen,
        onCheckedValueChange,
        checkedValues,
        defaultCheckedValues,
        hasMenuContext: true,
        triggerRef,
        onHover,
        onContext,
        triggerId,
      }}
    >
      {state.menuTrigger}
      {state.open && <slots.menuPopup {...slotProps.menuPopup} />}
    </MenuProvider>
  );
};
