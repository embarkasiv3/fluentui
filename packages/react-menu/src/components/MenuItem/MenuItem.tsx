import * as React from 'react';
import { useMenuItem } from './useMenuItem';
import { MenuItemProps } from './MenuItem.types';
import { renderMenuItem } from './renderMenuItem';
import { useMenuItemStyles } from './useMenuItemStyles';

/**
 * Define a styled MenuItem, using the `useMenuItem` and `useMenuItemStyles` hook.
 * {@docCategory MenuItem}
 */
export const MenuItem = React.forwardRef<HTMLElement, MenuItemProps>((props, ref) => {
  const state = useMenuItem(props, ref, {
    role: 'menuitem',
    tabIndex: 0, // TODO keyboard navigation
  });

  useMenuItemStyles(state);
  return renderMenuItem(state);
});

MenuItem.displayName = 'MenuItem';
