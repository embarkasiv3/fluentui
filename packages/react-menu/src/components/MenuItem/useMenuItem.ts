import * as React from 'react';
import {
  makeMergeProps,
  resolveShorthandProps,
  useMergedRefs,
  useEventCallback,
  shouldPreventDefaultOnKeyDown,
} from '@fluentui/react-utilities';
import { MenuItemProps, MenuItemState } from './MenuItem.types';
import { useCharacterSearch } from './useCharacterSearch';

/**
 * Consts listing which props are shorthand props.
 */
export const menuItemShorthandProps = ['icon'];

const mergeProps = makeMergeProps<MenuItemState>({ deepMerge: menuItemShorthandProps });

/**
 * Returns the props and state required to render the component
 */
export const useMenuItem = (
  props: MenuItemProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: MenuItemProps,
): MenuItemState => {
  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
      icon: { as: 'span' },
      role: 'menuitem',
      tabIndex: 0,
    },
    defaultProps,
    resolveShorthandProps(props, menuItemShorthandProps),
  );

  const { onKeyDown: onKeyDownOriginal } = state;
  state.onKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    if (shouldPreventDefaultOnKeyDown(e)) {
      e.preventDefault();
      (e.target as HTMLElement)?.click();
    }
    onKeyDownOriginal?.(e);
  };

  const { onMouseEnter: onMouseEnterOriginal } = state;
  state.onMouseEnter = useEventCallback((e: React.MouseEvent<HTMLElement>) => {
    state.ref.current?.focus();

    onMouseEnterOriginal?.(e);
  });

  useCharacterSearch(state);
  return state;
};
