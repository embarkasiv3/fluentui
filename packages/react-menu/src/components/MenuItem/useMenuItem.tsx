import * as React from 'react';
import {
  makeMergePropsCompat,
  resolveShorthandProps,
  useMergedRefs,
  useEventCallback,
  shouldPreventDefaultOnKeyDown,
} from '@fluentui/react-utilities';
import { MenuItemProps, MenuItemState } from './MenuItem.types';
import { useCharacterSearch } from './useCharacterSearch';
import { useMenuTriggerContext } from '../../contexts/menuTriggerContext';
import { ChevronRightIcon } from '../../utils/DefaultIcons';

/**
 * Consts listing which props are shorthand props.
 */
// TODO introduce content slot for styling
export const menuItemShorthandProps = ['icon', 'submenuIndicator', 'content', 'secondaryContent'] as const;

// eslint-disable-next-line deprecation/deprecation
const mergeProps = makeMergePropsCompat<MenuItemState>({ deepMerge: menuItemShorthandProps });

/**
 * Returns the props and state required to render the component
 */
export const useMenuItem = (
  props: MenuItemProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: MenuItemProps,
): MenuItemState => {
  const hasSubmenu = useMenuTriggerContext();

  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
      icon: { as: 'span' },
      submenuIndicator: { as: 'span', children: <ChevronRightIcon /> },
      content: { as: 'span', children: props.children },
      secondaryContent: { as: 'span' },
      role: 'menuitem',
      tabIndex: 0,
      hasSubmenu,
      'aria-disabled': props.disabled,
    },
    defaultProps,
    resolveShorthandProps(props, menuItemShorthandProps),
  );

  const { onClick: onClickOriginal, onKeyDown: onKeyDownOriginal } = state;
  state.onKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    if (shouldPreventDefaultOnKeyDown(e)) {
      if (state.disabled) {
        e.preventDefault();
        e.stopPropagation();
        return;
      }

      e.preventDefault();
      (e.target as HTMLElement)?.click();
    }

    onKeyDownOriginal?.(e);
  };

  state.onClick = (e: React.MouseEvent<HTMLElement>) => {
    if (state.disabled) {
      e.preventDefault();
      e.stopPropagation();
      return;
    }

    onClickOriginal?.(e);
  };

  const { onMouseEnter: onMouseEnterOriginal } = state;
  state.onMouseEnter = useEventCallback((e: React.MouseEvent<HTMLElement>) => {
    state.ref.current?.focus();

    onMouseEnterOriginal?.(e);
  });

  useCharacterSearch(state);
  return state;
};
