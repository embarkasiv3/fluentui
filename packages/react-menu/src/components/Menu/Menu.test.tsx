import * as React from 'react';
import { Menu } from './Menu';
import * as renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { MenuTrigger } from '../MenuTrigger/index';
import { MenuList } from '../MenuList/index';
import { MenuItem } from '../MenuItem/index';
import { MenuProps } from './Menu.types';

describe('Menu', () => {
  isConformant({
    disabledTests: [
      'as-renders-html',
      'as-renders-fc',
      'component-handles-ref',
      'component-has-root-ref',
      'component-handles-classname',
      'as-passes-as-value',
    ],
    Component: Menu,
    displayName: 'Menu',
    requiredProps: {
      children: [
        <MenuTrigger key="trigger">
          <button>MenuTrigger</button>
        </MenuTrigger>,
        <MenuList key="item">
          <MenuItem>Item</MenuItem>
        </MenuList>,
      ],
    },
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for Menu in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(
      <Menu>
        <MenuTrigger>
          <button>Menu trigger</button>
        </MenuTrigger>
        <MenuList>
          <MenuItem>Item</MenuItem>
        </MenuList>
      </Menu>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it.each([
    ['onMouseEnter', fireEvent.mouseEnter],
    ['onMouseLeave', fireEvent.mouseLeave],
    ['onBlur', fireEvent.blur],
    ['onKeyDown', fireEvent.keyDown],
  ])('should pass original %s handler to menu popup', (handler, trigger) => {
    // Arrange
    const spy = jest.fn();
    const menuPopup: MenuProps['menuPopup'] = { [handler]: spy };
    const { getByRole } = render(
      <Menu menuPopup={menuPopup}>
        <MenuTrigger>
          <button>Menu trigger</button>
        </MenuTrigger>
        <MenuList>
          <MenuItem>Item</MenuItem>
        </MenuList>
      </Menu>,
    );

    // Act
    fireEvent.click(getByRole('button'));
    trigger(getByRole('menu'));

    // Assert
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
