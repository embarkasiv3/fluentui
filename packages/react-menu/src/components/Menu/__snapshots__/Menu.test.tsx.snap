// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders a default state 1`] = `
<button
  aria-haspopup={true}
  id="1"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Menu trigger
</button>
`;
